pkg/sql/execinfrapb/processors.go

```diff
@@ -371,3 +380,90 @@ func (spec *WindowerSpec_Frame) ConvertToAST() (*tree.WindowFrame, error) {
 func (spec *JoinReaderSpec) IsIndexJoin() bool {
 	return len(spec.LookupColumns) == 0 && spec.LookupExpr.Empty()
 }
+
+// init performs some sanity checks for the invariants required by the
+// upperBuffer type.
+func init() {
+	isAllASCII := func(s string) bool {
+		for i := range s {
+			if s[i] >= utf8.RuneSelf {
+				return false
+			}
+		}
+		return true
+	}
+	// Check that aggregate function names are not longer than upperBufferSize
+	// and that they do not have non-ASCII characters. If these invariants
+	// change in the future, upperBufferSize or ToUpper will need to be
+	// adjusted.
+	for funcStr := range AggregatorSpec_Func_value {
+		if len(funcStr) > upperBufferSize {
+			panic(errors.AssertionFailedf(
+				"aggregate function name length cannot exceed length %d: %q",
+				upperBufferSize, funcStr,
+			))
+		}
+		if !isAllASCII(funcStr) {
+			panic(errors.AssertionFailedf(
+				"aggregate function name cannot contain non-ASCII characters: %q", funcStr,
+			))
+		}
+	}
+
+	// Perform the same check for window function names.
+	for funcStr := range WindowerSpec_WindowFunc_value {
+		if len(funcStr) > upperBufferSize {
+			panic(errors.AssertionFailedf(
+				"window function name length cannot exceed length %d: %q",
+				upperBufferSize, funcStr,
+			))
+		}
+		if !isAllASCII(funcStr) {
+			panic(errors.AssertionFailedf(
+				"window function name cannot contain non-ASCII characters: %q", funcStr,
+			))
+		}
+	}
+}
+
+const (
+	// upperBufferSize is large enough to accommodate the longest aggregate or
+	// window function name.
+	upperBufferSize = 31
+)
+
+// upperBuffer is a helper struct for creating a temporary upper-cased string
+// without performing a heap allocation. See ToUpper.
+type upperBuffer struct {
+	buf  [upperBufferSize]byte
+	used bool
+}
+
+// ToUpper returns a string where every lowercase ASCII character in "s" has
+// been converted to an uppercase character. If the length of "s" is greater
+// than upperBufferSize, ok=false is returned. If ToUpper returns ok=true, all
+// future invocations on the same upperBuffer will return ok=false.
+func (ub *upperBuffer) ToUpper(s string) (_ string, ok bool) {
+	if ub.used {
+		// Don't allow the buffer to be reused.
+		return "", false
+	}
+	if len(s) > upperBufferSize {
+		// The init function guarantees that no aggregate or window function has
+		// a name longer than upperBufferSize bytes. We can return ok=false here
+		// because upper-casing is pointless - the lookups in
+		// GetAggregateFuncIdx and GetWindowFuncIdx would fail anyway.
+		return "", false
+	}
+	ub.used = true
+	for i := 0; i < len(s); i++ {
+		c := s[i]
+		if c >= 'a' && c <= 'z' {
+			c -= 'a' - 'A'
+		}
+		ub.buf[i] = c
+	}
+	// Convert the buffer to a string without allocating.
+	b := ub.buf[:len(s)]
+	return *(*string)(unsafe.Pointer(&b)), true
```

Comment:
I'm really surprised this doesn't escape. I wasn't aware you could ever get the address of a stack pointer in Go. Taking the address of `b` should cause `ub` to be forced to the heap. What am I missing?

Nit: the unsafe package makes this somewhat more copacetic nowadays and not rely on the string header being a subset of a slice header. You can write this as `unsafe.String(&ub.buf[0], len(s))`.

Reply:
I was suprised too, but taking the address of an object doesn't automatically cause a heap allocation. You can take a look at the assembly here (I've removed the `errors` package to simplify things a bit): https://godbolt.org/z/e5c9TPMv8

The compiler is smart enough to realize that `&b` can point to the stack:

```
LEAQ    command-line-arguments.ub+32(SP), DX
MOVQ    DX, command-line-arguments.b+64(SP)
```

Here's a simpler example showing that you're not doomed to have a heap allocation for every `&` operator (although... maybe escape analysis doesn't even play a part here since the struct is just a single integer and the address-of operator is elided entirely):

```go
package foobar

type S struct {
    i int
}

func (s *S) addOne() {
    s.i++
}

func foo(x int) int {
    var s S
    s.i = x
    sp := &s
    sp.addOne()
    return s.i
}
```

```
TEXT    command-line-arguments.foo(SB), NOSPLIT|NOFRAME|ABIInternal, $0-8
FUNCDATA        $0, gclocals·g2BeySu+wFnoycgXfElmcg==(SB)
FUNCDATA        $1, gclocals·g2BeySu+wFnoycgXfElmcg==(SB)
FUNCDATA        $5, command-line-arguments.foo.arginfo1(SB)
FUNCDATA        $6, command-line-arguments.foo.argliveinfo(SB)
PCDATA  $3, $1
XCHGL   AX, AX
INCQ    AX
XCHGL   AX, AX
RET
```


> Nit: the unsafe package makes this somewhat more copacetic nowadays and not rely on the string header being a subset of a slice header. You can write this as unsafe.String(&ub.buf[0], len(s)).

Thanks for the tip. Done.

Reply:
TIL. Btw, today I learned about a `noescape` hack in the Go runtime: https://github.com/golang/go/blob/master/src/runtime/stubs.go#L126-L136. There is also this version which adds a `go:nocheckptr` declaration: https://github.com/golang/go/blob/master/src/strings/builder.go#L21-L32

It might be better to use this technique here as it seems to be used for a similar purpose: prevent stack allocated pointers from escaping to the heap.

Reply:
Do you thinking `noescape` adds value here when already have the assertion:

```
var ub upperBuffer //gcassert:noescape
```

Reply:
🤷🏼 The `gcassert:noescape` is certainly protecting you against this accidentally breaking. Yet the `noescape` function is how the Go runtime folks are ensuring a pointer doesn't escape. If you choose not to use `noescape` and the assert fires at some point, I'll definitely say "I told you so".

Reply:
Fair enough. Done.

Reply:
> I'm really surprised this doesn't escape. I wasn't aware you could ever get the address of a stack pointer in Go. Taking the address of b should cause ub to be forced to the heap. What am I missing?

I thought about this some more, and maybe this is less surprising when you think about this common pattern for reducing allocations:

```go
func foo() {
    var scratch [10]int
    s := appendUsuallyLessThanTenInts(scratch[:0])
    // ...
}

func appendUsuallyLessThanTenInts([]int) []int {
    // ...
}
```

The slice passed to `appendUsuallyLessThanTenInts` has a pointer to the stack-allocated `scratch` array.

Reply:
Hmm. Good point. I guess escape analysis is smart enough in both these cases to recognize that the pointer isn't actually escaping.